### PR TITLE
`this` should not be added to the scope bag

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Since we know how `this` works:

- we omit `this` from the `scope` bag always, as it's a syntax error to do `{ this }`
- we should still error if `this` shows up in a place where `this` doesn't exist (outside of classes and functions)
  - I couldn't figure out how to effectively do this that didn't have a ton of footguns, so I left it as an exercise for another time.
    For example, this is an easy use case where we want to forbid this:
    ```gts
    export default class Link extends Helper<Signature> {
      @service declare router: RouterService;
    
      compute(
        [href]: [href: string],
        { includeActiveQueryParams = false }: { includeActiveQueryParams?: boolean | string[] }
      ) {
        assert('href was not passed in', href);
    
        const isExternal = isExternal(href);
        const router = this.router;
        const handleClick = this.handleClick;
    
        // if this is accessed here -- we should error, because the `this` available
        // is not correct
        const Element = <template>
          <a
            data-active={{ (isActive router href includeActiveQueryParams) }}
            href={{if href href "##missing##"}}
            {{on "click" handleClick}}
            ...attributes
          >
            {{yield (hash isExternal=isExternal isActive=isActive)}}
          </a>
        </template>;
    
        return { 
          isExternal, 
          get isActive() {
            return isActive(router, href, includeActiveQueryParams);
          }, 
          Element 
        };
      }
    }
    ```
    This could maybe be written more simply in the underlying APIs, getting around the `this` restriction altogether by forcing folks to alias the variable -- which maybe they could do in the inline version above as well :thinking: 
    ```js
    compute(...) {
      return {
        Element: wireUpElement(this, href, includeActiveQueryParams)
         // ...
      }
    }
    // ...
    
    function wireUpElement(context, href, includeActiveQueryParams) {
      return setComponentTemplate(precompileTemplate(`
          <a
            data-active={{ (isActive context.router href includeActiveQueryParams) }}
            href={{if href href "##missing##"}}
            {{on "click" context.handleClick}}
            ...attributes
          >
            {{yield (hash isExternal=(isExternal href) isActive=isActive)}}
          </a>
      `, {
        context, on, href, includeActiveQueryParams
      }), templateOnly());
    }
    ```
    But this implies that we need to remove some caching that we have in ember right now -- because we don't want to keep component references around forever -- because `wireUpElement` won't care about previous component bindings next time it's called, even though the precompileTemplate part will be totally static, we're making a new binding between the scope bag and new templateOnly() instance.
    
    In any case, I think it warrants discussion about how to make this work ergonomically before restricting `this` usage arbitrarily 
    
   That's my morning, tangential, mind spew

### CODE REVIEW

File: src/scope-locals.ts [hunk 1]
```diff
@@ -79,6 +79,10 @@ export class ScopeLocals {
     return ['this', 'globalThis'].includes(jsName) || jsPath.scope.getBinding(jsName);
   }
 
+  #isImplicitScope(name: string) {
```
Comment (NullVoxPopuli) [inline src/scope-locals.ts:82]: the fix so far -- step next is to throw a hard error if `this` is referenced outside of a class or function

File: package.json [hunk 2]
```diff
@@ -86,7 +86,7 @@
     }
   },
   "volta": {
-    "node": "18.15.0",
+    "node": "14.21.3",
```
Comment (NullVoxPopuli) [inline package.json:89]:
```suggestion
    "node": "18.15.0",
```

**Comment by patricklx** [comment]
Is there another way to reference `this`. I can see how it would work for component templates.
But what about template in tests. shouldn't it possible to reference `this`. I guess we should still support it for wire format? And only if it doesn't have a component.
Not sure when hbs format is taking place. Is it happening as a step before wire or is it for addons?

**Comment by NullVoxPopuli** [comment]
>  But what about template in tests. shouldn't it possible to reference this. I guess we should still support it for wire format? And only if it doesn't have a component.

in tests, you can define this:
```js
test('...', function () {
  class Foo extends Component {
    hi = 'hi';
    
    <template>{{this.hi}}</template>
  }
}
```

which is valid.

**Comment by patricklx** [comment]
What about 

```gjs
test('...', function () {
  this.hi = b;
  tpl =  <template>{{this.hi}}</template>
}
```

**Comment by NullVoxPopuli** [comment]
I don't think that can be supported, setComponentTemplate would have to get a reference to that function first, and it's anonymous

**Comment by NullVoxPopuli** [comment]
> If this is passed with the scope it would work.

well, the transform currently does `scope: () => ({ this })` which is illegal, you'd have to
- rewrite all `this` references to `context` in the template
- change the `this` in the scope bag to `scope: () => ({ context: this })`

**Comment by patricklx** [comment]
> > If this is passed with the scope it would work.
> 
> well, the transform currently does `scope: () => ({ this })` which is illegal, you'd have to
> 
> * rewrite all `this` references to `context` in the template
> * change the `this` in the scope bag to `scope: () => ({ context: this })`

that is because its the hbs format.
in wire format it would be something like `scope: [this, other]`. So I think we could just disable adding `this` in hbs format

**Comment by ef4** [comment]
> What about
> 
> ```gjs
> test('...', function () {
>   this.hi = b;
>   tpl =  <template>{{this.hi}}</template>
> }
> ```

In the present design of components, it's intentional that that doesn't work. Because that's a template-only component and template-only components never have a `this`.

I don't think it's completely unreasonable to propose a change that would reclaim this unused piece of the namespace and make lexically-scoped `this` work. But I do think that would need some design discussion and at the very least a small amendment to the template-tag RFC. (Which is something we need to do anyway, since we've found enough cases that it's silent on.)

**Comment by ef4** [comment]
> * change the `this` in the scope bag to `scope: () => ({ context: this })`

No, if we decided to allow lexically scoped `this` is would just be `scope: () => ({ "this": this })`.

But also, there is existing API for how you pass `this` into the template https://github.com/emberjs/babel-plugin-ember-template-compilation/blob/ab7578e693bb0cd52c2add5cad6f94dd347cb8b7/__tests__/tests.ts#L1352

Disambiguating between these two and considering how it relates to the component manager API is why I'm saying this would require design work.

**Comment by patricklx** [comment]
> > * change the `this` in the scope bag to `scope: () => ({ context: this })`
> 
> No, if we decided to allow lexically scoped `this` is would just be `scope: () => ({ "this": this })`.
> 
> But also, there is existing API for how you pass `this` into the template
> 
> https://github.com/emberjs/babel-plugin-ember-template-compilation/blob/ab7578e693bb0cd52c2add5cad6f94dd347cb8b7/__tests__/tests.ts#L1352
> 
> Disambiguating between these two and considering how it relates to the component manager API is why I'm saying this would require design work.

That would be good.
Actually, it looks like it's different. There, the `this`  refers to the component prototype. Since it's inside static. So it cannot be used directly as this context.

**Comment by NullVoxPopuli** [comment]
@ef4 anything blocking this PR? 

we've have some great tangential discussion! :sweat_smile:

**Pull request merged**